src/combinators/mod.rs

```diff
@@ -964,6 +1029,36 @@ impl<'a> Skip for io::Cursor<&'a mut [u8]> {
     }
 }
 
+impl BackToTheBuffer for &mut [u8] {
+    fn reserve_write_use<Tmp, Gen: Fn(Self) -> GenResult<(Self, Tmp)>, Before: Fn(Self, Tmp) -> GenResult<Self>>(self, reserved: usize, gen: &Gen, before: &Before) -> GenResult<Self> {
+        let (res, buf) = self.split_at_mut(reserved);
+        let (buf, tmp) = gen(buf)?;
+        let res = before(res, tmp)?;
+        if !res.is_empty() {
+            return Err(GenError::BufferTooBig(res.len()));
+        }
+        Ok(buf)
+    }
+}
+
+#[cfg(feature = "std")]
+impl BackToTheBuffer for Vec<u8> {
+    fn reserve_write_use<Tmp, Gen: Fn(Self) -> GenResult<(Self, Tmp)>, Before: Fn(Self, Tmp) -> GenResult<Self>>(mut self, reserved: usize, gen: &Gen, before: &Before) -> GenResult<Self> {
+        let start_len = self.len();
+        self.reserve(reserved);
+        unsafe { self.set_len(start_len + reserved) };
+        let (mut vec, tmp) = gen(self)?;
+        let tmp_vec = before(Vec::new(), tmp)?;
+        let tmp_written = tmp_vec.len();
+        if tmp_written != reserved {
+            unsafe { self.set_len(start_len) };
+            return Err(GenError::BufferTooBig(reserved - tmp_written));
+        }
+        vec[start_len..(start_len + reserved)].copy_from_slice(&tmp_vec[..]);
```

Comment:
Maybe add a `// FIXME: Figure out how to do this correctly without copying` here in case a future reader of the code has an idea.